torch/csrc/jit/passes/create_autodiff_subgraphs.cpp

```diff
@@ -117,11 +145,167 @@ void CreateAutodiffSubgraphs(Block * block, size_t threshold, std::vector<Node*>
+        // (the body depends on the output of f, so Vertex 1 depends on Vertex 0).
+        // For each Use of y, we find an owning node of y that is a part of the
+        // dependency graph (in this case, the Vertex containing the if Node*)
+        // and then record the dependency.
+        auto * owning_node = use.user;
+        if (owning_node == block->return_node()) {
+          // The return node is not in the dag. Carry on.
+          continue;
+        }
+        while (true) {
+          auto search = node_to_vertex.find(owning_node);
+          if (search == node_to_vertex.end()) {
+            owning_node = owning_node->owningBlock()->owningNode();
+            JIT_ASSERT(owning_node != nullptr);
+            continue;
+          }
+          dag.addEdge(node_to_vertex[node], search->second);
+          break;
+        }
+      }
+    }
+  }
+  return dag;
+}
+
+static void find_differentiable_groups(
+    detail::DynamicDAG<Node*>& dep_graph,
+    size_t distance_threshold=64,
+    size_t producer_edge_threshold=16) {
+  // A Vertex contains a Node* or a differentiable group of Node*.
+  // Perform graph contraction on dep_graph: contract two vertices(x, y) if
+  // the following conditions hold:
+  // - x, y can be merged to form a differentiable group
+  // - the contraction would not invalidate the dag (it creates no cycles).
+
+  // Iterate in reverse topological order
+  int64_t ord = dep_graph.size() - 1;
+  for (int64_t ord = dep_graph.size() - 1; ord >= 0; --ord) {
+    if (!dep_graph.at(ord)) continue;
+
+    auto* consumer = dep_graph.at(ord).value();
+    if (!shouldConsiderForMerge(consumer)) continue;
+
+    // Iterate through consumer->in_edges() in reverse topological order
+    // Those should already be sorted by the DynamicDAG data structure.
+
+    bool changed = false;
+    for (auto it = consumer->in_edges().rbegin(); it != consumer->in_edges().rend(); ++it) {
+      auto * producer = *it;
+      // The distance threshold makes this algorithm "not optimal": it will miss
+      // some possible contraction opportunities, but it hopefully lets us:
+      // 1) preserve locality of tensors. We don't want to keep them alive for too long.
+      // 2) Help bound the computation complexity for contractEdge
+      if (consumer->ord - producer->ord > distance_threshold) continue;
+      // This also helps bound the complexity: contractEdge's complexity involves
+      // min(|out_edges(producer)|, |in_edges(consumer)|).
+      if (std::min(producer->out_edges().size(), consumer->in_edges().size()) >
+          producer_edge_threshold) continue;
+      if (!shouldConsiderForMerge(producer)) continue;
+
+      // If the edge contraction is successful, consumer->out_edges()
+      // may have changed so we break out of this loop.
+      if (dep_graph.contractEdge(producer, consumer)) {
+        changed = true;
+        break;
+      }
+    }
```

Comment:
The `changed` flag is really unnecessary. You only set it in one place, so you could have just as easily done `++ord`